storage/zynq7000-flash/flashdrv.c

```diff
@@ -347,58 +355,58 @@ static ssize_t _flashdrv_read(unsigned int id, addr_t offs, void *buff, size_t l
 }
 
 
+/* cache_readCb_t - wrapper for _flashdrv_read (*/
+static ssize_t _flashdrv_readCb(uint64_t offs, void *buff, size_t len, cache_devCtx_t *ctx)
+{
+	return _flashdrv_read(ctx->id, offs + ctx->start, buff, len);
+}
+
 /* Block device interface */
 
-static int _flashdrv_sync(unsigned int id)
+/* cache_writeCb_t */
+static ssize_t _flashdrv_writeCb(uint64_t offs, const void *buff, size_t len, cache_devCtx_t *ctx)
 {
 	uint8_t *src;
 	addr_t dst;
 	ssize_t res;
-	unsigned int i, pgNb;
+	unsigned int i, regID = ctx->id, pgNb;
 
 	size_t pageSz, sectSz;
 	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
 
-	if (fdrv_common.regs[id].buffPos == 0) {
-		return EOK;
-	}
 
-	sectSz = CFI_SIZE_SECTION(cfi->regs[id].size);
+	sectSz = CFI_SIZE_SECTION(cfi->regs[regID].size);
 	pageSz = CFI_SIZE_PAGE(cfi->pageSize);
 	pgNb = sectSz / pageSz;
 
-	res = _flashdrv_sectorErase(id, fdrv_common.regs[id].start + (sectSz * fdrv_common.regs[id].sectID));
-	if (res < 0) {
-		mutexUnlock(fdrv_common.regs[id].lock);
-		return res;
-	}
+	offs += ctx->start;
 
+	res = _flashdrv_sectorErase(regID, offs);
 
 	for (i = 0; i < pgNb; ++i) {
-		dst = fdrv_common.regs[id].start + fdrv_common.regs[id].sectID * sectSz + i * pageSz;
-		src = fdrv_common.regs[id].buff + i * pageSz;
+		dst = offs + i * pageSz;
+		src = (uint8_t *)buff + i * pageSz;
 
-		res = _flashdrv_pageProgram(id, dst, src, pageSz);
+		res = _flashdrv_pageProgram(regID, dst, src, pageSz);
 		if (res < 0) {
 			return res;
 		}
-	}
 
-	fdrv_common.regs[id].sectID = (uint32_t)-1;
-	fdrv_common.regs[id].buffPos = 0;
+		res = sectSz;
```

Comment:
It doesn't make sense to assign `sectSz` to res in each iteration in the loop. You can return `sectSz` at the end of the function.